ledger/transaction.ts

```diff
@@ -0,0 +1,294 @@
+  changeAddress: string,
+  changeValue: BigNumber,
+  inputUTXOs: UTXO[],
+  inputDerivation: TapBip32Derivation[] | undefined,
+  taprootScript: Buffer,
+  tapInternalKey: Buffer
+): Promise<Psbt> {
+  const btcNetwork = network === 'Mainnet' ? networks.bitcoin : networks.testnet;
+  const psbt = new Psbt({ network: btcNetwork });
+
+  for (const utxo of inputUTXOs) {
+    psbt.addInput({
+      hash: utxo.txid,
+      index: utxo.vout,
+      witnessUtxo: {
+        script: taprootScript,
+        value: utxo.value,
+      },
+      tapBip32Derivation: inputDerivation,
+      tapInternalKey,
+    });
+  }
+  recipients.forEach((value) => {
+    psbt.addOutputs([
+      {
+        address: value.address,
+        value: value.amountSats.toNumber(),
+      },
+    ]);
+  });
+
+  psbt.addOutputs([
+    {
+      address: changeAddress,
+      value: changeValue.toNumber(),
+    }]
+  );
+
+  return psbt;
+}
+
+/**
+ * This function is used to create a native segwit transaction for the ledger
+ * @param inputUTXOs - the selected input utxos
+ * @param inputDerivation - the derivation data for the sender address
+ * @returns the psbt without any signatures
+ * */
+export async function createMixedPsbt(
+  network: NetworkType,
+  recipients: Array<Recipient>,
+  changeAddress: string,
+  changeValue: BigNumber,
+  inputUTXOs: UTXO[],
+  inputDerivation: Bip32Derivation[] | undefined,
+  witnessScript: Buffer,
+  taprootInputDerivation: TapBip32Derivation[] | undefined,
+  taprootScript: Buffer,
+  tapInternalKey: Buffer,
+): Promise<Psbt> {
+  const btcNetwork = network === 'Mainnet' ? networks.bitcoin : networks.testnet;
+  const psbt = new Psbt({ network: btcNetwork });
+
+  const transactionMap = new Map<string, Buffer>();
+  for (const utxo of inputUTXOs) {
+    const txDataApiUrl = `${
+      network === 'Mainnet' ? MAINNET_BROADCAST_URI : TESTNET_BROADCAST_URI
+    }/${utxo.txid}/hex`;
+    const response = await axios.get(txDataApiUrl);
+    transactionMap.set(utxo.txid, Buffer.from(response.data, 'hex'));
+  }
+
+  const segwitInputs = inputUTXOs.filter(utxo => utxo?.address === changeAddress);
+  const taprootInputs = inputUTXOs.filter(utxo => !segwitInputs.some(input => utxo.txid === input.txid));
+
+  // Adding Taproot inputs
+  for (const utxo of taprootInputs) {
+    psbt.addInput({
+      hash: utxo.txid,
+      index: utxo.vout,
+      witnessUtxo: {
+        script: taprootScript,
+        value: utxo.value,
+      },
+      tapBip32Derivation: taprootInputDerivation,
+      tapInternalKey,
+    });
+  }
+
+  // Adding Segwit inputs
+  for (const utxo of segwitInputs) {
+    psbt.addInput({
+      hash: utxo.txid,
+      index: utxo.vout,
+      // both nonWitnessUtxo and witnessUtxo are required or the ledger displays warning message
+      witnessUtxo: {
+        script: witnessScript,
+        value: utxo.value,
+      },
+      nonWitnessUtxo: transactionMap.get(utxo.txid),
+      bip32Derivation: inputDerivation,
+    });
+  }
```

Comment:
Not required, but this could be made a lot more efficient with a single loop instead of 4+ like so:

```suggestion
// add inputs
for(const utxo of inputUTXOs) {
  if(utxo.address === changeAddress) {
    // add segwit UTXO
    psbt.addInput({
      hash: utxo.txid,
      index: utxo.vout,
      // both nonWitnessUtxo and witnessUtxo are required or the ledger displays warning message
      witnessUtxo: {
        script: witnessScript,
        value: utxo.value,
      },
      nonWitnessUtxo: transactionMap.get(utxo.txid),
      bip32Derivation: inputDerivation,
    });
  } else {
    // add taproot UTXO
    psbt.addInput({
      hash: utxo.txid,
      index: utxo.vout,
      witnessUtxo: {
        script: taprootScript,
        value: utxo.value,
      },
      tapBip32Derivation: taprootInputDerivation,
      tapInternalKey,
    });
    }
  }
```

Reply:
Totally agree, I updated it, thanks